core/payment-driver/erc20/src/erc20/faucet.rs

```diff
@@ -0,0 +1,172 @@
+const DEFAULT_ETH_FAUCET_HOST: &str = "faucet.testnet.golem.network";
+const FAUCET_ADDR_ENVAR: &str = "ETH_FAUCET_ADDRESS";
+const MAX_FAUCET_REQUESTS: u32 = 6;
+
+lazy_static! {
+    static ref MIN_GLM_BALANCE: U256 = utils::big_dec_to_u256(BigDecimal::from(50));
+    static ref MIN_ETH_BALANCE: U256 = utils::big_dec_to_u256(BigDecimal::from_f64(0.005).unwrap());
+    static ref MAX_WAIT: Duration = Duration::minutes(1);
+}
+
+pub async fn request_glm(
+    dao: &Erc20Dao,
+    address: H160,
+    network: Network,
+) -> Result<(), GenericError> {
+    let str_addr = format!("0x{:x}", address);
+    let balance = ethereum::get_balance(address, network).await?;
+    if balance >= *MIN_ETH_BALANCE {
+        log::info!("Enough tETH balance.");
+    } else {
+        log::info!(
+            "Requesting tETH from erc20 faucet... address = {}",
+            &str_addr
+        );
+
+        for i in 0..MAX_FAUCET_REQUESTS {
+            match faucet_donate(address, network).await {
+                Ok(()) => break,
+                Err(e) => {
+                    // Do not warn nor sleep at the last try.
+                    if i >= MAX_FAUCET_REQUESTS - 1 {
+                        log::error!(
+                            "Failed to request tGLM from Faucet, tried {} times.: {:?}",
+                            MAX_FAUCET_REQUESTS,
+                            e
+                        );
+                        return Err(e);
+                    } else {
+                        log::warn!(
+                            "Retrying ({}/{}) to request tGLM from Faucet after failure: {:?}",
+                            i + 1,
+                            MAX_FAUCET_REQUESTS,
+                            e
+                        );
+                        delay_for(time::Duration::from_secs(10)).await;
+                    }
+                }
+            }
+        }
+        wait_for_eth(address, network).await?;
+    }
+    let glm_balance = ethereum::get_glm_balance(address, network).await?;
+
+    if glm_balance >= *MIN_GLM_BALANCE {
+        log::info!("Enough tGLM balance.");
+        return Ok(());
+    }
+    let pending = dao.get_pending_faucet_txs(&str_addr, network).await;
+    for _tx in pending {
+        log::info!("Already pending a mint transactin.");
+        return Ok(());
+    }
+    log::info!(
+        "Requesting tGLM from erc20 faucet... address = {}",
+        &str_addr
+    );
+
+    let network_nonce = wallet::get_network_nonce(address, network).await?;
+    let db_nonce = dao.get_next_nonce(&str_addr, network).await?;
+    let nonce = if network_nonce > db_nonce {
+        network_nonce
+    } else {
+        db_nonce
+    };
+
+    let db_tx = ethereum::sign_faucet_tx(address, network, nonce).await?;
+    // After inserting into the database, the tx will get send by the send_payments job
+    dao.insert_raw_transaction(db_tx).await;
+
+    delay_for(time::Duration::from_secs(10)).await;
```

Comment:
Why the delay?

Reply:
because it is very unlikely to be mined before then:
- cron runs every 5 seconds, so 2.5s avg
- blocks are minted every 15 seconds, so 7.5s avg

will add a comment for our future selves

Reply:
There's already sleeping in the `wait_for_glm` function but if we want to spare 3 calls to `ethereum::get_balance` then I guess leaving this is fine.